src/core/mesh/qgsmeshdataprovider.h

```diff
@@ -73,14 +145,70 @@ class CORE_EXPORT QgsMeshSource SIP_ABSTRACT
 
 /**
  * \ingroup core
-  * Base class for providing data for QgsMeshLayer
-  *
-  * Responsible for reading native mesh data
-  *
-  * \see QgsMeshSource
-  * \since QGIS 3.2
-  */
-class CORE_EXPORT QgsMeshDataProvider: public QgsDataProvider, public QgsMeshSource
+ * Dataset is a  collection of vector or scalar values on vertices or faces of the mesh
+ *
+ * Base on the underlying data provider/format, whole dataset is either stored in memory or
+ * read on demand
+ *
+ * \note The API is considered EXPERIMENTAL and can be changed without a notice
+ *
+ * \since QGIS 3.2
+ */
+class CORE_EXPORT QgsMeshDatasetSource SIP_ABSTRACT
+{
+  public:
+    //! Dtor
+    virtual ~QgsMeshDatasetSource() = default;
+
+    /**
+     * \brief Associate dataset with the mesh
+     */
+    virtual bool addDataset( const QString &uri ) = 0;
+
+    /**
+     * \brief Returns number of datasets loaded
+     */
+    virtual int datasetCount() const = 0;
+
+    /**
+     * \brief Returns whether dataset has scalar data associated
+     */
+    virtual bool datasetHasScalarData( int index ) const = 0;
+
+    /**
+     * \brief Returns whether dataset is on vertices
+     */
+    virtual bool datasetIsOnVertices( int index ) const = 0;
+
+    /**
+     * \brief Returns dataset metadata
+     */
+    virtual QgsMeshDatasetMetadata datasetMetadata( int index ) const = 0;
+
+    /**
+     * \brief Returns value associated with the index from the dataset
+     */
+    virtual QgsMeshDatasetValue datasetValue( int datasetIndex, int valueIndex ) const = 0;
+
+    /**
+     * \brief Returns whether dataset is valid
+     */
+    virtual bool datasetIsValid( int index ) const = 0;
```

Comment:
Can we merge datasetHasScalarData(), datasetIsOnVertices(), datasetMetadata(), datasetIsValid() into one function?

I would suggest to have one method datasetMetadata() that returns QgsMeshDatasetMetadata which would be a class with methods hasScalarData(), isOnVertices(), isValid() and extraOptions().